ENGINE: yse_tests_channel concurrency loop flaky hang/segfault in full ctest (spurious CI failures)

**Layer:** engine · **Category:** Concurrency / test-flakiness · **RT/audio-thread path?** No (control-plane teardown/concurrency, but it blocks CI)

### Summary

The `yse_tests_channel` target intermittently **hangs or segfaults** in the full **un-sanitized** `ctest --test-dir build` run. It is flaky (heap-/timing-dependent), so green CI runs are partly luck, and a bad roll produces a spurious PR failure. This is the same family as the (closed) SOUND/CHANNEL manager races #41 and the static-teardown UAF #298 — but it still reproduces on current `dev`, so either #298's fix is incomplete or a distinct race remains.

### Concrete repro — CI

PR #335 (commit `ce361fd`, C-API instrument surface — touches only `YseEngine/c_api/`, nothing in the channel path) failed its first "Build and analyze" job purely on this:

```
1/27 Test #1: yse_unit_tests ......... Passed 11.98 sec
 ... #2–#6 pass fast ...
       Start  7: yse_tests_channel
23:32:12  Start 7: yse_tests_channel
23:49:58  ##[error]The operation was canceled.      # ~18 min, no progress → job timeout
          Terminate orphan process: pid 4650 (ctest)
          Terminate orphan process: pid 44858 (yse_tests)
```

Run: actions/runs/28905924873 (job "Build and analyze"). A **re-run of the identical commit passed** the same target in 6m53s — confirming flakiness, not a code regression in #335. The sanitizer (asan/tsan) and Python presets pass because they don't exercise this target the same way; the failure is specific to the full non-sanitized `ctest`.

### Concrete repro — local (the decisive bit)

Running the channel-concurrency loop locally on **unmodified `dev` (a71dd5e, i.e. WITHOUT #335)** segfaults roughly **~20–40% of runs** — so the defect is pre-existing, not introduced by the instrument work. A binary-layout perturbation (any unrelated change to `libyse`) merely shifts the trip rate, which is why an innocent PR can flip a normally-green suite red.

### Why it matters

- Spurious red on unrelated PRs (cost us a full investigation + CI re-run on #335).
- Masks a real teardown/concurrency UAF: on CI it manifests as an 18-min hang (lost-wakeup / lock-order), locally as a SIGSEGV (use-after-free) — both point at the same race.

### Suggested next steps

- Reproduce under TSan/ASan **on the full non-sanitized target's scenario** (the sanitizer presets currently skip it), stress-looping `yse_tests_channel` (e.g. `ctest -R yse_tests_channel --repeat until-fail:50`).
- Audit the CHANNEL::Manager / SOUND impl teardown + the channel concurrency stress case against the #41 (manager-list race) and #298 (sound-impl-outlives-channel static-teardown UAF) fixes — check whether #298's coordinated-destruction fix covers the channel-concurrency path or left a gap.
- Consider isolating the concurrency stress case into its own binary so a hang can't stall the whole ctest job (mirrors the "isolate sensitive suites" note from #298/#304).

### Related

- #41 — SOUND::Manager unsynchronised cross-thread list access (closed, fixed)
- #298 — static-teardown UAF when a sound impl outlives its channel (closed, fixed) — **appears incomplete: crashes still reproduce on current dev**
- #304 — sensitive-suite isolation

--- a/Tests/channel/test_channel_concurrency.cpp
+++ b/Tests/channel/test_channel_concurrency.cpp
@@ -20,10 +20,13 @@
 #include <doctest/doctest.h>
 #include <atomic>
 #include <chrono>
+#include <cstring>
+#include <new>
 #include <string>
 #include <thread>
 #include "yse.hpp"
 #include "channel/channelInterface.hpp"
+#include "channel/channelImplementation.h"
 #include "channel/channelManager.h"
 #include "sound/soundManager.h"
 #include "internal/time.h"
@@ -45,6 +48,36 @@ namespace {
 } // namespace
 
 TEST_SUITE("channel") {
+
+  // ─── Regression: a freshly constructed impl is never seen as deletable ───────
+
+  TEST_CASE("channel concurrency: fresh impl status is OBJECT_CONSTRUCTED "
+            "regardless of prior memory (issue #336)") {
+    // Root cause of the flaky channel two-thread churn hang/segfault: the
+    // channel implementationObject ctor left `objectStatus` (a C++17
+    // std::atomic, whose default ctor does NOT zero the value) uninitialised.
+    // channel::create() emplaces the impl into the mutex-guarded
+    // `implementations` list (addImplementation) and only AFTERWARDS calls
+    // Manager().setup(), which sets OBJECT_CREATED. During that window the
+    // slow-pool deleteJob runs `implementations.remove_if(canBeDeleted)` with
+    // canBeDeleted == (objectStatus == OBJECT_DELETE). If the recycled heap slot
+    // happened to hold the OBJECT_DELETE bit-pattern, the job freed the node the
+    // worker thread was still constructing — a heap-layout-dependent UAF (crash)
+    // or corrupted intrusive list (hang). SOUND/REVERB already initialised
+    // objectStatus in their ctors; CHANNEL was the outlier.
+    //
+    // Poison the storage, then placement-construct an impl over it: with the ctor
+    // fix objectStatus reads OBJECT_CONSTRUCTED; without it the poison survives
+    // and the impl presents as some other (possibly deletable) state. This is
+    // deterministic — it fails on the unpatched ctor and passes with the fix.
+    alignas(YSE::CHANNEL::implementationObject) unsigned char
+        storage[sizeof(YSE::CHANNEL::implementationObject)];
+    std::memset(storage, 0xFF, sizeof(storage));
+    auto* impl = new (storage) YSE::CHANNEL::implementationObject(nullptr);
+    CHECK(impl->getStatus() == YSE::OBJECT_CONSTRUCTED);
+    CHECK_FALSE(YSE::CHANNEL::implementationObject::canBeDeleted(*impl));
+    impl->~implementationObject();
+  }
 
   // ─── Single-thread churn: many channel create→destroy cycles ─────────────────
 

--- a/YseEngine/channel/channelImplementation.cpp
+++ b/YseEngine/channel/channelImplementation.cpp
@@ -12,6 +12,19 @@
 
 YSE::CHANNEL::implementationObject::implementationObject(channel* head)
   : head(head),
+    // Initialise objectStatus in the ctor, not left to the first setStatus():
+    // addImplementation() emplaces this impl into the mutex-guarded
+    // `implementations` list BEFORE create() calls Manager().setup() (which sets
+    // OBJECT_CREATED). In C++17 std::atomic's default ctor leaves the value
+    // indeterminate, so during that window the slow-pool deleteJob —
+    // `implementations.remove_if(canBeDeleted)`, canBeDeleted == (status ==
+    // OBJECT_DELETE) — could read the garbage bytes as OBJECT_DELETE and free
+    // the node the worker thread is still constructing (heap-layout-dependent
+    // UAF / hang, issue #336). emplace_front runs under implementationsMutex and
+    // the setup/delete jobs read under the same mutex, so this OBJECT_CONSTRUCTED
+    // write is published before either job can observe the impl. SOUND and REVERB
+    // already initialise objectStatus this way; CHANNEL was the outlier.
+    objectStatus(OBJECT_CONSTRUCTED),
     newVolume(1.f),
     lastVolume(1.f),
     parent(nullptr),

